### PR TITLE
Inherit ContextDependentPiecewiseLinearConstraint in *Constraint classes

### PR DESCRIPTION
**Pull request by AleksandarZeljic**
This PR changes ReLU, AbsolutValue, Sign, Max and Disjunction Constraints 
to inherit ContextDependentPiecewiseLinearConstraint class, which prepares
them for transition to the new architecture.

- Implements virtual methods of ContextDependentPiecewiseLinearConstrain.
- Adds basic tests for context dependent behavior of all the constraint classes.
- Minor changes in Max and Disjunction constraint to use PhaseStatus enum.

Co-authored-by: ahmed-irfan <irfan@cs.stanford.edu>
Co-authored-by: anwu1219 <haozewu@stanford.edu>

### CODE REVIEW

File: src/engine/DisjunctionConstraint.cpp [hunk 1]
```diff
@@ -19,9 +19,24 @@
 #include "Statistics.h"
 
 DisjunctionConstraint::DisjunctionConstraint( const List<PiecewiseLinearCaseSplit> &disjuncts )
-    : _disjuncts( disjuncts )
-    , _feasibleDisjuncts( disjuncts )
+    : ContextDependentPiecewiseLinearConstraint( disjuncts.size() )
+    , _disjuncts( disjuncts.begin(), disjuncts.end() )
+    , _feasibleDisjuncts( disjuncts.size(), 0 )
```
Comment (guykatzz) [inline src/engine/DisjunctionConstraint.cpp:24]: why not just _feasibleDisjuncts( disjuncts )? Instead of the loop in the body of the function

File: src/engine/DisjunctionConstraint.cpp [hunk 2]
```diff
@@ -19,9 +19,24 @@
...
+    for ( unsigned ind = 0;  ind < disjuncts.size();  ++ind )
+        _feasibleDisjuncts.append( ind );
+
+    extractParticipatingVariables();
+}
+
+DisjunctionConstraint::DisjunctionConstraint( const Vector<PiecewiseLinearCaseSplit> &disjuncts )
+    : ContextDependentPiecewiseLinearConstraint( disjuncts.size() )
+    , _disjuncts( disjuncts )
+    , _feasibleDisjuncts( disjuncts.size(), 0 )
```
Comment (guykatzz) [inline src/engine/DisjunctionConstraint.cpp:35]: Same as above

File: src/engine/DisjunctionConstraint.cpp [hunk 3]
```diff
@@ -127,17 +150,35 @@ List<PiecewiseLinearConstraint::Fix> DisjunctionConstraint::getSmartFixes( ITabl
 
 List<PiecewiseLinearCaseSplit> DisjunctionConstraint::getCaseSplits() const
 {
-    return _disjuncts;
+    return List<PiecewiseLinearCaseSplit>( _disjuncts.begin(), _disjuncts.end() );
```
Comment (guykatzz) [inline src/engine/DisjunctionConstraint.cpp:153]: What's the difference between the line erased and the new line?

File: src/engine/DisjunctionConstraint.h [hunk 4]
```diff
@@ -174,15 +201,22 @@ class DisjunctionConstraint : public PiecewiseLinearConstraint
...
+        return static_cast<PhaseStatus>( ind + 1 );
+    }
+
+    inline unsigned phaseStatusToInd( PhaseStatus phase ) const
+    {
+        //ASSERT( phase != PHASE_NOT_FIXED );
+        return static_cast<unsigned>( phase ) - 1;
+    }
+
+
```
Comment (guykatzz) [inline src/engine/DisjunctionConstraint.h:218]: remove empty lines

File: src/engine/MaxConstraint.cpp [hunk 5]
```diff
@@ -196,7 +221,16 @@ void MaxConstraint::notifyUpperBound( unsigned variable, double value )
...
-        _elements.erase( variable );
+    {
+        if ( _cdInfeasibleCases )
+            markInfeasible( variableToPhase( variable ) );
+        else
+            _elements.erase( variable );
+    }
+
+    // If all elements have been eliminated, set the phase of the constraint.
+    if ( _elements.size() == 0)
```
Comment (guykatzz) [inline src/engine/MaxConstraint.cpp:232]: use _elements.empty()

File: src/engine/MaxConstraint.h [hunk 6]
```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
...
-    String serializeToString() const;
+    String serializeToString() const override;
 
     /*
      * Returns a boolean value indicating if at least one input variable was eliminated (True)
      * or not (False)
      */
     bool wereVariablesEliminated() const;
 
+
```
Comment (guykatzz) [inline src/engine/MaxConstraint.h:174]: remove empty lines

File: src/engine/MaxConstraint.h [hunk 7]
```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
...
     bool _obsolete;
     bool _eliminatedVariables;
     double _maxValueOfEliminated;
 
+    /*
+       Functions that abstract away _maxIndex and _maxIndexSet, and use the
+       refactored PhaseStatus to store this information.
+
+       Makes two assumptions:
+        - 0 is not a valid variable value, since it's reserved for PHASE_NOT_FIXED
```
Comment (guykatzz) [inline src/engine/MaxConstraint.h:193]: 0 is actually almost always a variable. can we get rid of this assumption?

File: src/engine/MaxConstraint.h [hunk 8]
```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
...
+        - the total number of variables does not reach the value of MAX_PHASE_ELIMINATED
+     */
+    inline bool maxIndexSet() const
+    {
+        return getPhaseStatus() != PHASE_NOT_FIXED;
```
Comment (guykatzz) [inline src/engine/MaxConstraint.h:198]: I realized I don't fully understand the semantics of _phaseStatus and _cdPhaseStatus. Is this an indication of the phase that a non-fixed constraint is in? For MAX this makes sense (indicates the current element with maximal assignment), but what does it mean for Relu and Abs?

File: src/engine/MaxConstraint.h [hunk 9]
```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
...
+    inline PhaseStatus variableToPhase( unsigned variable ) const
+    {
+        return static_cast<PhaseStatus>( variable );
+    }
+
+    inline unsigned phaseToVariable( PhaseStatus phase ) const
+    {
+        return static_cast<unsigned>( phase );
+    }
+
```
Comment (guykatzz) [inline src/engine/MaxConstraint.h:237]: Remove empty lines

File: src/engine/ReluConstraint.cpp [hunk 10]
```diff
@@ -31,7 +33,8 @@
 #endif
 
 ReluConstraint::ReluConstraint( unsigned b, unsigned f )
-    : _b( b )
+    : ContextDependentPiecewiseLinearConstraint( 2u )
```
Comment (guykatzz) [inline src/engine/ReluConstraint.cpp:36]: Use constant

File: src/engine/SignConstraint.cpp [hunk 11]
```diff
@@ -27,7 +29,8 @@
 #endif
 
 SignConstraint::SignConstraint( unsigned b, unsigned f )
-    : _b( b )
+    : ContextDependentPiecewiseLinearConstraint( 2u )
```
Comment (guykatzz) [inline src/engine/SignConstraint.cpp:32]: Constant

**Review by guykatzz — COMMENTED** [review]
Most of my comments are cosmetic. This is a pretty major change - have you run all available benchmarks to see that nothing breaks?

File: src/engine/AbsoluteValueConstraint.cpp [hunk 12]
```diff
@@ -71,17 +73,25 @@ PiecewiseLinearFunctionType AbsoluteValueConstraint::getType() const
     return PiecewiseLinearFunctionType::ABSOLUTE_VALUE;
 }
 
-PiecewiseLinearConstraint *AbsoluteValueConstraint::duplicateConstraint() const
+ContextDependentPiecewiseLinearConstraint *AbsoluteValueConstraint::duplicateConstraint() const
 {
     AbsoluteValueConstraint *clone = new AbsoluteValueConstraint( _b, _f );
     *clone = *this;
+    this->initializeDuplicatesCDOs( clone );
```
[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.cpp:80]: Fixed

File: src/engine/AbsoluteValueConstraint.cpp [hunk 13]
```diff
@@ -329,6 +339,21 @@ List<PiecewiseLinearCaseSplit> AbsoluteValueConstraint::getCaseSplits() const
     return splits;
 }
 
+List<PhaseStatus> AbsoluteValueConstraint::getAllCases() const
+{
+    return { ABS_PHASE_NEGATIVE, ABS_PHASE_POSITIVE};
```
[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.cpp:344]: PhaseStatus is a shared ENUM between all ConstraintClasses. I can add the constant with some generic name - BIPHASIC_PIECEWISE_LINEAR_CONSTRAINT. Although, it would make sense to have it as standalone constant to avoid mixing up status and number of cases. And it would be only constant of its kind, since all constraints so far are either fixed biphasic or have dynamic number of cases. What do you think?

File: src/engine/AbsoluteValueConstraint.cpp [hunk 14]
```diff
@@ -378,7 +403,7 @@ bool AbsoluteValueConstraint::phaseFixed() const
     return _phaseStatus != PhaseStatus::PHASE_NOT_FIXED;
 }
 
-PiecewiseLinearCaseSplit AbsoluteValueConstraint::getValidCaseSplit() const
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getImpliedCaseSplit() const
```
[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.cpp:406]: As we move to a new SMT engine, we will switch to terms decision and impliedCase, which will return PhaseStatus cases and be used to retrieve the CaseSplit itself. So this method will go away soon anyway, I renamed it to indicate that they serve the same purpose.

File: src/engine/AbsoluteValueConstraint.h [hunk 15]
```diff
@@ -2,23 +2,33 @@
...
- ** [[ Add lengthier description here ]]
-
+ ** AbsoluteValueConstraint implements the following constraint:
+ ** f = Abs( b )
+ **
+ ** It distinguishes two relevant phases for search:
+ ** ABS_PHASE_POSITIVE: b > 0
+ ** ABS_PHASE_NEGATIVE: b <=0
+ **
+ ** The constraint operates in two modes pre-processing mode, which is stores
```
[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.h:19]: Fixed.

File: src/engine/DisjunctionConstraint.cpp (hunk 1 above)
Comment (AleksandarZeljic) [reply src/engine/DisjunctionConstraint.cpp:24]: The data structure differ now, it used to be a List and now its a Vector to utilize indexing using PhaseStatus. I didn't want to meddle with outer interface for now, but if you want I can go fix that where needed.

File: src/engine/DisjunctionConstraint.cpp (hunk 3 above)
Comment (AleksandarZeljic) [reply src/engine/DisjunctionConstraint.cpp:153]: Return type _disjuncts is a Vector now. This method will become obsolete in the long run.

File: src/engine/MaxConstraint.h (hunk 7 above)
Comment (AleksandarZeljic) [reply src/engine/MaxConstraint.h:193]: Yeah, I introduced the conversion functions since, it is simple to just increment the stored value by one when reading/writing.

File: src/engine/MaxConstraint.h (hunk 8 above)
Comment (AleksandarZeljic) [reply src/engine/MaxConstraint.h:198]: In my mind the semantics is the same, the only difference is when we can act on the information. ReLU and other two-phase constraints do not depend on any additional variables, so we can treat them as implications immediately. Max is trickier because we need to convince ourselves that no other variable can exceed the value, before we can treat it as an implication.

File: src/engine/AbsoluteValueConstraint.cpp [hunk 16]
```diff
@@ -71,17 +73,25 @@ PiecewiseLinearFunctionType AbsoluteValueConstraint::getType() const
...
     *clone = *this;
+    this->initializeDuplicatesCDOs( clone );
     return clone;
 }
 
 void AbsoluteValueConstraint::restoreState( const PiecewiseLinearConstraint *state )
 {
     const AbsoluteValueConstraint *abs = dynamic_cast<const AbsoluteValueConstraint *>( state );
+
+    CVC4::context::CDO<bool> *activeStatus = _cdConstraintActive;
```
[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.cpp:88]: So this is not fantastic logic, it is there simply to avoid losing context objects and causing memory leaks. Once we switch to context-dependent architecture the restoreState methods all over the place will be obsolete.

[the remark this answers is not in the document]
Comment (AleksandarZeljic) [reply src/engine/AbsoluteValueConstraint.cpp:88]: I can re-initialize their values to whatever we are restoring to, but it is really just paranoid code and not likely to be relevant in practice. Alternatively, I can replace current logic with assertions to indicate that the constraint is not in context-dependent mode.

File: src/engine/MaxConstraint.h (hunk 8 above)
Comment (guykatzz) [reply src/engine/MaxConstraint.h:199]: Sorry, I still don't follow. If max's phaseStatus() is x1, does this mean that currently x1 has the highest assignment?
Are we "overloading" this field to indicate this?

Comment (AleksandarZeljic) [reply src/engine/MaxConstraint.h:199]: Yes, x1 would have the highest assignment, or higher lower bound than all the other upper bounfs. 

I don't think the field is overloaded. 
Relu = max ( x, 0) vs Max = max( x, y). 

Relu can propagate as soon as we know something about x since zero is a constant. But max(x,y) can propagate only once it also knows that the upper bound of y is less than the value of x, otherwise it is not sufficient information. 

Basically, _phaseStatus can be thought of as the phase under the current bounds (and implicitly assignment). Relu needs information over one variable for propagation, but max needs also to know bounds on all the other variables involved to propagate. _phaseStatus denotes the same, the difference is that ReLU never has to wait on other variables so there is no period of uncertainty. Does this make sense?

**Comment by AleksandarZeljic** [comment]
@guykatzz I had to re-base the branch, can you approve again? There was one bug fix since your last review.
